### ISSUE TITLE
Refresh token session handling has a race condition

### ISSUE THREAD

**Issue by chancez** [oauth2-proxy/oauth2-proxy#1992]
## Expected Behavior

I've described the issue in depth here: https://github.com/oauth2-proxy/oauth2-proxy/issues/1473#issuecomment-1407108808 but I'll repeat it here:

When trying to figure out why oauth2 proxy isn't working with Gitlab, after trying various changes to figure out what's missing when authenticating to gitlab, I discovered, it's actually working, but the newly refreshed session is being invalidated due to what seems to be some kind of race condition.

I added more logging into the OIDC provider and the oauth2 package, and from what I can tell, there's a race condition when two requests occur at the same time and the ID token is expired and needs refreshing. I see two refresh attempts happening concurrently, one fails, one succeeds. The one that fails is likely failing because the refresh token is revoked when the other refresh succeeds. The failing refresh tries to verify the ID token which is expired, and then deletes the session/cookie because it's no longer valid (see this `session is invalid, removing session` log).

You'll notice two requests failing, that's because the underlying OAuth2 library retries automatically if the first request fails, because it tries first with the clientID and clientSecret in the query parameters, then if that fails, it tries again with the clientID and clientSecret in the basic auth headers.

Here's the logs:

```
[2023/01/27 21:37:52] [stored_session.go:193] Refreshing session - User: 778521; SessionAge: 3m54.69892219s
vencode: grant_type=refresh_token&refresh_token=omitted
req url https://gitlab.com/oauth/token
[2023/01/27 21:37:52] [stored_session.go:193] Refreshing session - User: 778521; SessionAge: 3m54.69892219s
vencode: grant_type=refresh_token&refresh_token=omitted
req url https://gitlab.com/oauth/token
finished redeem, err failed to get token: oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}
[2023/01/27 21:37:53] [stored_session.go:197] Unable to refresh session: error refreshing tokens: unable to redeem refresh token: failed to get token: oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}
[2023/01/27 21:37:53] [oidc.go:107] id_token verification failed: failed to verify token: oidc: token is expired (Token Expiry: 2023-01-27 21:35:55 +0000 UTC)
[2023/01/27 21:37:53] [stored_session.go:98] Error loading cookied session: error refreshing access token for session (Session{email:chance.zibolski@gmail.com user:778521 PreferredUsername:chancez token:true id_token:true created:2023-01-27 21:33:57.30107781 +0000 UTC expires:2023-01-27 23:33:55.83307877 +0000 UTC refresh_token:true}): session is invalid, removing session
127.0.0.1:54656 - dc1f9206ea961e01a040a1efd7d8b8d1 - - [2023/01/27 21:37:52] hubble-ui.192-168-105-101.sslip.io GET - "/oauth2/auth" HTTP/1.0 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0" 401 13 1.006
127.0.0.1:54666 - dc1f9206ea961e01a040a1efd7d8b8d1 - - [2023/01/27 21:37:53] hubble-ui.192-168-105-101.sslip.io GET - "/oauth2/sign_in" HTTP/1.0 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0" 200 8374 0.000
got new token refreshed &{omitted Bearer omitted 2023-01-27 23:37:54.10517144 +0000 UTC m=+7481.721031418 map[access_token:omitted created_at:1.674855473e+09 expires_in:7200 id_token:omitted.omitted.omitted refresh_token:omitted scope:openid email profile token_type:Bearer]}
finished redeem, err <nil>
```

I'm pretty sure this is the cause of the issues because I've got a oauth2 CLI project that has no issues using very similar logic to what's being used in oauth2 proxy. I've compared the urls, the request bodies, headers, the OAuth2 flows, and everything seems to match up. The CLI works, OAuth2 proxy does not.

## Steps to Reproduce (for bugs)

Add logging here: https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/oidc.go#L147 and then try using gitlab with refresh enabled. You'll see logs similar what I had above. One succeeds, another doesn't, and the user is logged out.

## Context

Basically, it breaks using gitlab with OIDC because ID tokens have a 2 minute lifetime, making it effectively required to use refresh tokens.

## Your Environment

I'm running it in a local kubernetes cluster in KIND, but it shouldn't matter.

- Version used: 7.4.0/master

**Comment by miguelborges99** [comment]
Are you using cookie based storage or redis based storage? Oauth2-proxy is stateless, so it is hard to know if there is another parallel session doing a refresh at the same time. The state is kept in a cookie or in redis, depending on your configuration. In a cookie based storage, it is not possible to guarantee 100% that only one request is able to refresh the session, i.e. recreate the session cookie. You may have two or more Oauth2-proxy instances, and no locks in the code would solve that. We could try to create another cookie to signal that refresh is ongoing, but what would this solve? Parallel requests could happen in the same milisecond and could fail to see this cookie. 
I see that in redis based storage there are some locks, maybe with this configuration you might have more luck.

**Comment by chancez** [comment]
This in cookie based sessions. The way I've done something similar is to use `golang.org/x/sync/singleflight`. But yeah, there needs to be some form of locking to on the actual mechanism to refresh the token. This is basically required because OIDC spec does say it's valid for the IDP to revoke existing tokens when you issue a new refresh token. Without this, any provider that does such revocation on refresh will likely not work properly.

What's interesting is there is already usage of atomics in oauth2-proxy's session code to avoid concurrency bugs when updating the internal session objects, but it's only locking the critical section of updating the local in-memory session, not locking on the actual refresh on the session itself.

**Comment by JoelSpeed** [comment]
The locks for the session state are only for the single request, there's no cross request locking. We saw this race condition many times in the past and that's why we came up with the server side session store and abstracted the locking into the session storage. If you want to protect against this, my recommendation is to use the redis session store, or, implement another session storage that is persistent.

The alternative, which can work but isn't guaranteed, is to use, as you say, something like singlelifght. I've used nginx with a short caching to prevent multiple requests from being sent to OAuth2 Proxy at once, but it relies on sticky sessions and as I say, isn't 100% reliable

Server side session storage is the path forward

**Comment by chancez** [comment]
I mean fundamentally this means cookie based session storage is broken. Moving to a separate session storage merely punts the locking problem somewhere else. I think it makes sense for some installations, but I'd wager the average oauth2 proxy configuration is far below that threshold, so I think fixing the cookie based storage makes sense. 

Currently this only affects some IDPs, but you're at the whims of the providers and their configuration for refresh tokens.

> The alternative, which can work but isn't guaranteed, is to use, as you say, something like singlelifght. 

Why wouldn't it be guaranteed to work?

**Comment by JoelSpeed** [comment]
> Why wouldn't it be guaranteed to work?

Because you need to guarantee that all requests from a single client hit the same OAuth2 Proxy backend. So either, you have a single replica of OAuth2 Proxy, or you use some sort of sticky sessions in front to make sure that you always hit the same instance, which can be tricky in many setups
